packages/hardhat/contracts/strategy/CrabOTC.sol

```diff
@@ -0,0 +1,207 @@
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {EIP712} from "@openzeppelin/contracts/drafts/EIP712.sol";
+import {ECDSA} from "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import {StrategyMath} from "./base/StrategyMath.sol";
+
+contract CrabOTC is ReentrancyGuard, EIP712 {
+    using StrategyMath for uint256;
+    using Address for address payable;
+
+    address public immutable crab;
+    address public immutable weth;
+    address public immutable controller;
+    address public immutable wPowerPerp;
+    mapping(address => mapping(uint256 => bool)) public nonces;
+
+    struct Order {
+        address initiator;
+        address trader;
+        uint256 quantity;
+        uint256 price;
+        bool isBuying;
+        uint256 expiry;
+        uint256 nonce;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    event DepositOTC(
+        address indexed depositor,
+        uint256 crabAmount,
+        uint256 wSqueethAmount,
+        uint256 depositedAmount,
+        uint256 executedPrice,
+        address trader
+    );
+
+    event WithdrawOTC(
+        address indexed depositor,
+        uint256 crabAmount,
+        uint256 ethSent,
+        uint256 wSqueethAmount,
+        uint256 executedPrice,
+        address trader
+    );
+
+    bytes32 private constant _CRAB_BALANCE_TYPEHASH =
+        keccak256("Order(address initiator,address trader,uint256 quantity,uint256 price,bool isBuying,uint256 expiry,uint256 nonce)");
+
+    constructor(address _crab) EIP712("CrabOTC", "2") {
+        require(_crab != address(0), "Invalid crab address");
+
+        crab = _crab;
+        weth = ICrabStrategyV2(_crab).weth();
+        controller = ICrabStrategyV2(_crab).powerTokenController();
+        wPowerPerp = ICrabStrategyV2(_crab).wPowerPerp();
+    }
+
+    /**
+     * @notice set nonce to true
+     * @param _nonce the number to be set true
+     */
+    function setNonceTrue(uint256 _nonce) external {
+        nonces[msg.sender][_nonce] = true;
+    }
+
+    function deposit(
+        uint256 _totalEth,
+        uint256 _minPrice,
+        Order memory _order
+    ) external payable nonReentrant {
+        _verifyOrder(_order, true, _minPrice);
+        uint256 depositedEth = msg.value;
+
+        uint256 wSqueethQuantity = _getWSqueethToMint(_totalEth);
+        require(_order.quantity >= wSqueethQuantity, "Order quantity is less than needed");
+
+        _order.quantity = wSqueethQuantity;
+
+        uint256 neededEth = _totalEth.sub(depositedEth);
+        uint256 wethAmount = _order.quantity.wmul(_order.price);
+        require(wethAmount >= neededEth, "Need more ETH");
+
+        IWETH9(weth).transferFrom(_order.trader, address(this), wethAmount);
+        IWETH9(weth).withdraw(wethAmount);
+
+        ICrabStrategyV2(crab).deposit{value: _totalEth}();
+        uint256 crabAmount = IERC20(crab).balanceOf(address(this));
+        IERC20(crab).transfer(_order.initiator, crabAmount);
```

Comment:
So in this case the order signer should know who they are buying oSQTH from before hand, right?
Can't we just use msg.sender here? and keep it flexible based on who execute this order

Reply:
So the problem is anyone with the order can execute the trade, and getting the order from ipfs is not that difficult. This is just to make sure the intended person executes the order